src/library/scala/collection/immutable/List.scala

```diff
@@ -395,6 +397,19 @@ sealed abstract class List[+A]
     None
   }
 
+  override def last: A = {
+    if (isEmpty) throw new NoSuchElementException("List.last")
+    else {
+      var these = this
```

Comment:
Perhaps you already tried, but I think you can do this loop with a single variable:
```Scala
var curr = this
while (curr.tail != Nil) curr = curr.tail 
curr.head
```

I am not sure if the Scala compiler would be clever enough to optimise several access to a same field, given that there are no field writes on the loop.

Reply:
I copy/pasted the implemenetation from the superclass. Up there, it probably tries to avoid calling `.tail` more than once because its an abstract method and its performance characteristics (and even side effects) are not known.